### ISSUE TITLE
no training results

### ISSUE THREAD

**Issue by anandkoirala** [yhenon/keras-frcnn#35]
I am working on windows
- couldn't install tensorflow for python 2.7 and without it i couldn't train using train_frcnn.py
- so i used python 3.5 and fixed all the errors 
- training started displays "Starting training" but nothing happens......

**Comment by prajwal967** [comment]
What are the errors that we need to fix for running it on python3.5?

**Comment by anandkoirala** [comment]
Testing is working but training has gone in a continuous loop... if someone has a fix let me know
Search opencv installation folder for cv2.pyd  copy this file inside site-packages folder of your python installation
go to cmd and type python
inside python console type import cv2 
if error you have to install opencv3 version: go to http://www.lfd.uci.edu/~gohlke/pythonlibs/ and download your appropriate wheel from Opencv section
paste the downloaded file to your working directory of cmd prompt
now open cmd and do pip install downloaded_file_name_with_extension.whl
success? follow the steps:

1. convert python2 file to python3
the easiest way to fix syntax is to use 2to3.py which is inside scripts folder of python. open cmd and cd to the directory where your file myfile.py is located and then type this in the cmd prompt
python path_to_2to3.py -w myfile.py
2. install h5py 
pip install h5py
3. download trained model_frcnn.hdf5 and config.pickle file for the same model and save it in the working directory.  I cd inside the keras-frcnn folder so i pasted there
https://drive.google.com/open?id=0B2XPZBSClSzhNmlMb3AzNW5YVW8
https://drive.google.com/open?id=0B2XPZBSClSzhdENkRmM1alVsN3c
4.Try to run using the command 
python test_frcnn.py -p /path/to/test_data/
test_data can be a folder containing your images

Error Sections:
1. If scalar value for augmented 'color' not numeric
- test_frcnn line 205 change class_to_color to (0,255,0)

**Comment by anandkoirala** [comment]
Hi Prajwal,
Comments made on the website
Regards,
Anand

**Comment by anandkoirala** [comment]
@yhenon i have changed files to work in python 3.5.2 and started training...  However sometimes it throws exception as 'float object cannot be interpreted as int' but the program still runs...  There were syntax changes made on float and int types in python3.. if we can make sure which expression has to return what data type this can be fixed. 
please reply me and maybe we would be able to fix and upload the codes for python 3..
Regards
Anand

**Comment by anandkoirala** [comment]
@yhenon 'exception float cannot be interpreted as int'  should be fixed...  I have trained despite this error and working good..  However for some test images the bounding box is drawn outside of image boundary which could be because of the data type conversion error...
![bus1](https://cloud.githubusercontent.com/assets/22046591/26391864/c6b4aa7c-40a7-11e7-8522-c4866fc2d259.PNG)
![bus](https://cloud.githubusercontent.com/assets/22046591/26391878/ce423fd4-40a7-11e7-9203-5a5b1f9064fe.PNG)
![bus3](https://cloud.githubusercontent.com/assets/22046591/26391954/28f801c0-40a8-11e7-9e64-b2725843e8a9.PNG)

**Comment by anandkoirala** [comment]
trained for bus and car only

**Comment by yhenon** [comment]
Hi @anandkoirala , I'd definitely be interested in integrating the python3 support. I'll try and set up a python3 env for myself to test this.

I'm surprised at the low quality of the bounding boxes you're getting for that bus image given how low the loss for the regression layer is in your training, is it common?

**Comment by anandkoirala** [comment]
thanks @yhenon 
yes maybe because i trained with only 7 images each of car and bus.
after the training was completed below is the results
![bus6](https://cloud.githubusercontent.com/assets/22046591/26429307/794279b8-4129-11e7-91bd-d046cdd81300.PNG)
![bus4](https://cloud.githubusercontent.com/assets/22046591/26429317/7c95d36c-4129-11e7-9fbe-0b12ea287401.PNG)
![bus5](https://cloud.githubusercontent.com/assets/22046591/26429318/80623530-4129-11e7-9562-7a7a926b0074.PNG)

**Comment by anandkoirala** [comment]
@yhenon following were the steps i took ....  and i am eagerly waiting for the working python3.5 version from you.
- pip install future
- futurize --stage1 file_to_be_converted.py
- futurize --stage2 file_to_be_converted.py
above will make files compatible with python2 and python3 
**train_frcnn.py**
- codes for opening and saving files: replace 'r' and 'w' with 'rb' and 'wb'
- change data_gen_train.next() to next(data_gen_train)
**data_generators.py**
- remove the line all_img_data = sorted(all_img_data)
- y_rpn_regr[:, y_rpn_regr.shape[1]//2:, :, :] *= C.std_scaling   double backslash (without which i was not able to train as the generator did not yield.. this is syntax change in python3 for float)

**Comment by tokpanov** [comment]
I was also able to convert the code to Python3.5 through futurize (with some minor code changes). I really don't understand the reason for the line all_img_data = sorted(all_img_data) in data_generators.py. Anyways, this dataset is shuffled. Why sort it then?

**Comment by anandkoirala** [comment]
(someone correct me if i am wrong)...
@tokpanov 
There was no key used to sort the dictionary and as the data was shuffled so I removed the line all_img_data = sorted(all_img_data) without any noticeable changes on the output...

**Comment by piyush-J** [comment]
@anandkoirala 
I followed the steps as you had mentioned above. However, I am encountering this problem:
*Additional Info*
My keras.json file contains:
{
    "epsilon": 1e-07, 
    "floatx": "float32", 
    "image_data_format": "channels_first", 
    "backend": "tensorflow"
}
I would be glad if you could please help me out. 
![capture](https://user-images.githubusercontent.com/29255684/26883095-d2335dd6-4bb9-11e7-9d31-3539b7ac73e5.PNG)

**Comment by anandkoirala** [comment]
@piyush-J  
keras.json   channels_last for tensorflow and channels_first for theano
change this and see if that works

**Comment by tokpanov** [comment]
@anandkoirala Yes, I trained and tested on our own dataset and it works.

**Comment by piyush-J** [comment]
@anandkoirala 
I had tried that out too, but it showed the same error message.

**Comment by anandkoirala** [comment]
@piyush-J  try these
1. Try upgrading keras and tensorflow  and test
2. follow the steps in my comments for testing (start from line 2 if you have already futurize) and test on one of pascal voc images to make sure the code is working...   I have gone through this problem in my initial stages but can't make sure for the fix as my computer is frozen for training at this moment..

**Comment by piyush-J** [comment]
Thank you very much @anandkoirala. I used the futurize method to convert my codes and after some minor code modifications, the testing part is working fine.

**Comment by yhenon** [comment]
If someone wants to submit a PR with changes for python 3 it would help me
out

**Comment by piyush-J** [comment]
Dear Mr. Yann,

I would be glad to help you. I just need some time because I have few tight
deadlines at my workplace. I will contribute once I get a little free after
a week or so.

Regards,
Piyush Jha

**Comment by Ahleroy** [comment]
I have done a pull request in order to use the Faster R-CNN with Python 3.5 [#70](https://github.com/yhenon/keras-frcnn/pull/70)

**Comment by Poojadj** [comment]
When I test my model the image is rescaled, would wish to evaluate the predictions against the ground truth bounding boxes. Meaning that because the image is rescaled when entered into the model the bounding boxes are also rescaled and therefore I can't compare them to the original ground truth because image scale difference. Is there a way to resolve this?

**Comment by anandkoirala** [comment]
@Poojadj  I am not a python programmer but what i can tell from my understanding is...  if you go to data_genetators.py line 125-129  where the groundtruth annotations are resized with a ratio depending on the original size of individual image. the best option would be to add some line of code in that section to output those data to a new file which would serve as a new groundtruth for evaluation .....
cheers

**Comment by Ahleroy** [comment]
The test_frcnn.py has been modified in the pull request #70 . The program now saves the original image with the bounding boxes at the right dimensions.

Regards ,

**Comment by HuangBo-Terraloupe** [comment]
I also got it works for python3, but something strange is the losses starts different when I using python3 and python2. 
For python3, the regression loss of RPN starts with 7, but for python2 it is only 1.5.

**Comment by anandkoirala** [comment]
@HuangBo-Terraloupe  i have noticed 7 sometimes and 4 many times but i think it should not matter.  however if the value is different for two version of python for the same data set something looks wrong.
I am working on a custom dataset which is completely different from any of the image set available over internet. The objects are really small in size so if i used the default IoU value of 0.5 i would expect big number as starting RPN loss.. changing IoU to 0.2 for my dataset should work for my object size and I expect low value of starting RPN loss. However if I keep training it would go lower to about 0.02

**Comment by HuangBo-Terraloupe** [comment]
@anandkoirala 
Thanks for your answer,
I also work on a custom dataset, and I also need to detect small things.
Do you have some suggestions about how to detect small things?
One thing I come to in my mind is use smaller anchor size, 
My image patches are all 500x500, and the object are really small, maybe 20x20.

**Comment by anandkoirala** [comment]
@HuangBo-Terraloupe 
this comment is strictly based on personal experiences however suggestions and comments are welcomed...
good to hear back from you..  you have to be clear that the objects in images have to be at least 16x16pixels when image is around 500x500 pixels that is because of 16 pix stride implementation of fasterRCNN for final layer feature map and model downsampling (u don't want to miss objects). the other thing is the method of IoU struggles as the object size decreases (more error for smaller objects). luckily u have 20x20 in 500x500 for me it is even worse so i have to split images to get object greater than 16x16 pix. there are papers if you google 'company logo detection' which explains the failure of Faster-RCNN for small objects the main problem is with object localization.. even worse with YOLO ( whic is what you want for real-time object detection with less accuracy in localization) some approach use the feature map from initial layers or concatenate from different layers for small objects.. for your case yes you have to change anchor sizes ( don't have to follow power of 2) depending on the size variation of objects in images and also the aspect ratio to match how your object looks like..  the other suggestion would be to change the IoU thresh ( i mean when >0.5 it is +ve , <0.1 is -ve) 0.5 to 0.2 or 0.3.  this is to do with localization.. there are times strictly u need to know your dataset... sometimes you don't even have to go very deep in neural networks a simple LeNet can do your job or a combination of cascade classifier backed by CNN..  i don't really know about how the losses are implemented in this repo ( something different to official faster_RCNN?). your losses doesn't look that bad as the model struggles for object localization not classification for small objects.. however comparatively the first two losses should have gone low for better results..  you can shoot an email to me a.koirala@cqu.edu.au in detail for further discussions... also may be we could be of some help to each other....
